Src/DependencyCollector/Shared/Implementation/SqlClientDiagnosticSourceListener.cs

```diff
@@ -0,0 +1,539 @@
+            
+            this.client.Track(telemetry);
+        }
+
+        private void OnCommandErrorEvent(
+            Guid operationId,
+            string operation,
+            Guid connectionId,
+            SqlCommand command,
+            Exception exception,
+            long endTimestamp)
+        {
+            var telemetry = this.CreateCommandTelemetry(operationId, command, endTimestamp);
+
+            ConfigureExceptionTelemetry(telemetry, exception);
+
+            this.client.Track(telemetry);
+        }
+
+        private DependencyTelemetry CreateCommandTelemetry(Guid operationId, SqlCommand command, long endTimestamp)
+        {
+            // Call DeriveDuration first to ensure we remove any start timestamp from operationStartTimestamps
+            var duration = this.DeriveDuration(operationId, endTimestamp);
+
+            var dependencyName = string.Empty;
+            var target = string.Empty;
+
+            if (command.Connection != null)
+            {
+                target = string.Join(" | ", command.Connection.DataSource, command.Connection.Database);
+
+                var commandName = command.CommandType == CommandType.StoredProcedure
+                    ? command.CommandText
+                    : string.Empty;
+
+                dependencyName = string.IsNullOrEmpty(commandName)
+                    ? string.Join(" | ", command.Connection.DataSource, command.Connection.Database)
+                    : string.Join(" | ", command.Connection.DataSource, command.Connection.Database, commandName);
+            }
+
+            var telemetry = new DependencyTelemetry()
+            {
+                Id = operationId.ToString("N"),
+                Name = dependencyName,
+                Type = RemoteDependencyConstants.SQL,
+                Target = target,
+                Data = command.CommandText,
+                Duration = duration,
+                Timestamp = DateTimeOffset.UtcNow - duration,
+                Success = true
+            };
+
+            InitializeTelemetry(telemetry, operationId);
+
+            return telemetry;
+        }
+
+        private void OnErrorConnectionEvent(
+            Guid operationId,
+            string operation,
+            Guid connectionId,
+            SqlConnection connection,
+            Exception exception,
+            long endTimestamp)
+        {
+            var duration = this.DeriveDuration(operationId, endTimestamp);
+
+            var telemetry = new DependencyTelemetry()
+            {
+                Id = operationId.ToString("N"),
+                Name = string.Join(" | ", connection.DataSource, connection.Database, operation),
+                Type = RemoteDependencyConstants.SQL,
+                Target = string.Join(" | ", connection.DataSource, connection.Database),
+                Data = operation,
+                Duration = duration,
+                Timestamp = DateTimeOffset.UtcNow - duration,
```

Comment:
We anyway store the precise timestamp when operation started, so we should just use that, instead of deriving, which wont be as accurate. (There is a UI showing dependency timeline, which could get affected if start time is not as precise as possible, esp. for fast calls.)

Reply:
We are not storing the start timestamp, but the ticks value returned by Stopwatch.GetTimestamp().

Reply:
ok